File: objective-c/ArcGISSample/Sketch/SketchViewController.m
//
//  SketchViewController.m
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

#import "SketchViewController.h"


@implementation SketchViewController


- (void)viewDidLoad
{
    
    [super viewDidLoad];
    
    AGSMapView *agsMapView = [[AGSMapView alloc] initWithFrame:self.view.bounds];
	[self.view addSubview:agsMapView];
    
    //タイルマップサービスレイヤーの追加
    NSURL *url = [NSURL URLWithString:@"https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer"];
    AGSTiledMapServiceLayer *agsTiledMapServiceLayer = [[AGSTiledMapServiceLayer alloc] initWithURL:url];
    [agsMapView addMapLayer:agsTiledMapServiceLayer withName:@"Tiled Layer"];
    
    //スケッチしたグラフィック表示用のグラフィックスレイヤーの追加
    self.agsGraphicsLayer = [AGSGraphicsLayer graphicsLayer];
    [agsMapView addMapLayer:self.agsGraphicsLayer withName:@"Graphics Layer"];
    
    //スケッチレイヤーの追加
    self.agsSketchGraphicsLayer = [[AGSSketchGraphicsLayer alloc] initWithGeometry:nil];
    self.agsSketchGraphicsLayer.midVertexSymbol = nil;
    [agsMapView addMapLayer:self.agsSketchGraphicsLayer withName:@"Sketch Layer"];
    
    //作成するジオメトリタイプの指定
    AGSMutablePolygon *agsSketchGeom = [[AGSMutablePolygon alloc] init];
    self.agsSketchGraphicsLayer.geometry = agsSketchGeom;
    
    agsMapView.touchDelegate = self.agsSketchGraphicsLayer;
    
    
    UIBarButtonItem *buttonUndo = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemUndo target:self action:@selector(undoSketch)];
    UIBarButtonItem *buttonRedo = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemRedo target:self action:@selector(redoSketch)];
    UIBarButtonItem *buttonRemove = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemTrash target:self action:@selector(removeVertex)];
    UIBarButtonItem *buttonSubmit = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemSave target:self action:@selector(submitFeature)];
    UIBarButtonItem *flexibleItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
    NSArray *buttons = [NSArray arrayWithObjects:buttonUndo, flexibleItem, buttonRedo, flexibleItem, buttonRemove, flexibleItem, buttonSubmit, nil];

	UIToolbar *toolbar = [[UIToolbar alloc] initWithFrame:CGRectMake(0, self.view.frame.size.height - 44, self.view.frame.size.width, 44)];
	[toolbar setItems:buttons];	
	[self.view addSubview:toolbar];
    

}


- (void)undoSketch {
    
    //編集を元に戻す
    if ([self.agsSketchGraphicsLayer.undoManager canUndo]) {
        [self.agsSketchGraphicsLayer.undoManager undo];
    }
    
}

- (void)redoSketch {
    
    //編集をやり直す
    if ([self.agsSketchGraphicsLayer.undoManager canRedo]) {
        [self.agsSketchGraphicsLayer.undoManager redo];
    }
    
}

- (void)removeVertex {
    
    //選択されている頂点を削除する
    [self.agsSketchGraphicsLayer removeSelectedVertex];
    
}

- (void)submitFeature {
    
    //作成したジオメトリからシンボルを指定してグラフィックを作成し、グラフィックスレイヤーに追加
    AGSSimpleFillSymbol *fillSymbol = [AGSSimpleFillSymbol simpleFillSymbol];
    fillSymbol.color = [[UIColor purpleColor] colorWithAlphaComponent:0.25];
    fillSymbol.outline.color = [UIColor darkGrayColor];
    
    AGSGraphic *agsGraphic = [[AGSGraphic alloc] initWithGeometry:self.agsSketchGraphicsLayer.geometry symbol:fillSymbol attributes:nil];
    [self.agsGraphicsLayer addGraphic:agsGraphic];
    
    [self.agsSketchGraphicsLayer clear];

}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



@end

File: swift/ArcGISSample/SketchViewController.swift
//
//  SketchViewController.swift
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

import UIKit
import ArcGIS


class SketchViewController: UIViewController{
    
    
    var agsGraphicsLayer: AGSGraphicsLayer!
    var agsSketchGraphicsLayer: AGSSketchGraphicsLayer!
    
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        let agsMapView = AGSMapView(frame: view.bounds)
        view.addSubview(agsMapView)
        
        //タイルマップサービスレイヤーの追加
        let url = URL(string: "https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer")
        let tiledLyr = AGSTiledMapServiceLayer(url:url)
        agsMapView.addMapLayer(tiledLyr, withName:"Tiled Layer")
        
        //スケッチしたグラフィック表示用のグラフィックスレイヤーの追加
        agsGraphicsLayer = AGSGraphicsLayer()
        agsMapView.addMapLayer(agsGraphicsLayer, withName:"Graphics Layer")
        
        //スケッチレイヤーの追加
        agsSketchGraphicsLayer = AGSSketchGraphicsLayer()
        agsSketchGraphicsLayer.midVertexSymbol = nil
        agsMapView.addMapLayer(agsSketchGraphicsLayer, withName:"Sketch Layer")
        
        //作成するジオメトリタイプの指定
        agsSketchGraphicsLayer.geometry = AGSMutablePolygon()
        agsMapView.touchDelegate = agsSketchGraphicsLayer
        
        
        let buttonUndo = UIBarButtonItem(barButtonSystemItem: .undo, target: self, action: #selector(SketchViewController.undoSketch))
        let buttonRedo = UIBarButtonItem(barButtonSystemItem: .redo, target: self, action: #selector(SketchViewController.redoSketch))
        let buttonRemove = UIBarButtonItem(barButtonSystemItem: .trash, target: self, action: #selector(SketchViewController.removeSketch))
        let buttonSubmit = UIBarButtonItem(barButtonSystemItem: .save, target: self, action: #selector(SketchViewController.submitSketch))
        let flexibleItem = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)


        let buttons = ([buttonUndo, flexibleItem, buttonRedo, flexibleItem, buttonRemove, flexibleItem, buttonSubmit])
        let toolbar: UIToolbar = UIToolbar(frame: CGRect(x: 0, y: view.frame.size.height - 44, width: view.frame.size.width, height: 44))
        toolbar.setItems(buttons as [UIBarButtonItem], animated: true)
        view .addSubview(toolbar)
        
    }
    
    
    func undoSketch(_ sender: UIBarButtonItem) {
        
        //編集を元に戻す
        if (agsSketchGraphicsLayer.undoManager.canUndo) {
            agsSketchGraphicsLayer.undoManager.undo()
        }
        
    }
    
    func redoSketch(_ sender: UIBarButtonItem) {
        
        //編集をやり直す
        if (agsSketchGraphicsLayer.undoManager.canRedo) {
            agsSketchGraphicsLayer.undoManager.redo()
        }
        
    }
    
    func removeSketch(_ sender: UIBarButtonItem) {
        
        //選択されている頂点を削除する
        agsSketchGraphicsLayer.removeSelectedVertex()
        
    }
    
    
    func submitSketch(_ sender: UIBarButtonItem) {
        
        //作成したジオメトリからシンボルを指定してグラフィックを作成し、グラフィックスレイヤーに追加
        let fillSymbol = AGSSimpleFillSymbol(color: UIColor.purple .withAlphaComponent(0.25), outlineColor: UIColor.darkGray)
        let agsGraphic = AGSGraphic(geometry: agsSketchGraphicsLayer.geometry, symbol: fillSymbol, attributes: nil)
        
        agsGraphicsLayer.addGraphic(agsGraphic)
        agsSketchGraphicsLayer.clear()
        
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}
